### ISSUE TITLE
`yas-snippet-dirs` does not update when package.el installs a new version

### ISSUE THREAD

**Issue by vermiculus** [capitaomorte/yasnippet#495]
Consider yasnippet vA to be loaded. Package management gives it a timestamp on the directory, such as `yasnippetA/`. Loading yasnippet now adds the installation directory to the variable.

Update the package to version B; now the installation directory is `yasnippetB`. Yasnippet now complains because `yasnippetA` no longer exists.

Is there a way to work around this issue?

**Comment by kootenpv** [comment]
I do not understand what the current fix is from this?

Both these do not work:

    (yas/load-directory (concat emacsd "snippets"))
    (setq yas-snippet-dirs (concat emacsd "snippets"))

**Comment by npostavs** [comment]
If you don't do anything, it should already just work. If you want only your own snippets, you can use

```elisp
(setq yas-snippet-dirs '("~/.emacs.d/snippets"))
```

**Comment by kootenpv** [comment]
@npostavs Sorry for my unclear message, I meant to say that it is still loading the package's snippets as well. I want those disabled (for now I deleted the snippets folder, but I suppose those would be back on a yasnippet upgrade)

**Comment by npostavs** [comment]
@kootenpv Okay, `(setq yas-snippet-dirs '("~/.emacs.d/snippets"))` should work for that, but you have to set it *before* running `(yas-global-mode +1)`, or `(yas-reload-all)`.

**Comment by kootenpv** [comment]
Indeed, I had `(yas-global-mode t)` before `yas-snippet-dirs`, that was the issue, thanks.

**Comment by computersarecool** [comment]
This does not make any sense to me.

I continually get the error `Opening directory: no such file or directory, /home/user/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode`

This is because in that folder I only have `yasnippet-20161221.1953` which is a newer version.

I do not even want those older snippets, in fact all I want is my own snippets, but even when I define `(setq yas-snippet-dirs '("~/.emacs.d/snippets"))` in my `init.el` I still get this error.

My packages are:

package-archives '(("gnu" . "http://elpa.gnu.org/packages/") 
                               ("melpa" . "http://melpa.milkbox.net/packages/")  
                                 ("marmalade" . "http://marmalade-repo.org/packages/")))


The readme says: `Yasnippet no longer bundles snippets directly` but clearly it is because I am getting this error.

How can I fix this?

**Comment by npostavs** [comment]
@computersarecool Do you get anything from `M-x list-load-path-shadows`?

> The readme says: `Yasnippet no longer bundles snippets directly` but clearly it is because I am getting this error.

The snippets are not in the git repository, but they are in the MELPA package (via the subrepo).

**Comment by computersarecool** [comment]
This is the output:
```
/usr/share/emacs/24.5/site-lisp/debian-startup hides /usr/share/emacs/site-lisp/debian-startup
1 Emacs Lisp load-path shadowing was found  
```

Is there any way to not get those snippets from the git repository / MELPA package? 
I can't even run my `init.el` file like this...

It seems like the file in ELPA has a newer timestamp and no matter what I can't just use my own snippets.

**Comment by npostavs** [comment]
Hmm, well I don't understand where the incorrect setting is coming from. Maybe try uninstalling and reinstalling yasnippet?

Otherwise you could install it manually from git. https://github.com/joaotavora/yasnippet#install-the-most-recent-version

**Comment by computersarecool** [comment]
I have tried doing this about 4 times today.  In fact I have uninstall all my ELPA packages.

Do you know any possible way that this could be searching for a folder with a different timestamp?

**Comment by npostavs** [comment]
What's the value of `yas-installed-snippets-dir`? I suppose it's pointing to `~/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode`? Maybe you can bisect your config to find out who sets it?

**Comment by computersarecool** [comment]
It is indeed `"/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets"  `

I have no idea what sets it because the only code I have in my .init file, related to YAS is:
```
(require 'yasnippet)       
(setq yas-snippet-dirs '("~/.emacs.d/snippets"))      
(yas-global-mode 1)
```

**Comment by computersarecool** [comment]
There is also a call to yasnippet in my editor config which is set by `custom` but I doubt that has anything to do with it...

**Comment by npostavs** [comment]
Well, it might not be written in your init.el specifically, but in `emacs -Q` `yas-installed-snippet-dirs` is not set, so *something* in their must be setting it...

**Comment by computersarecool** [comment]
Is there a way to overwrite this somehow so that it only checks my `snippets` directory (which I made).  And what else is weird is I am never even **calling** css mode or a css file.

**Comment by npostavs** [comment]
> Is there a way to overwrite this somehow so that it only checks my snippets directory

`(setq yas-snippet-dirs '("~/.emacs.d/snippets"))` should do it, but you've said you already have that.

Maybe a backtrace would be informative. Try running adding `(setq debug-on-error t)` at the top of your init file. Or if the error occurs during startup, start Emacs with `emacs --debug-init`.

**Comment by computersarecool** [comment]
The debug error I get is:


```
Debugger entered--Lisp error: (file-error "Opening directory" "no such file or directory" "/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode")                                      
  directory-files("/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode" t)                                                                                                            
  yas--subdirs("/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode" no-subdirs-just-files)                                                                                           
  yas--load-directory-2("/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode" css-mode)                                                                                               
  yas--load-directory-1("/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode" css-mode)                                                                                               
  (lambda nil (yas--load-directory-1 (quote "/home/optonox/.emacs.d/elpa/yasnippet-20161201.1520/snippets/css-mode") (quote css-mode)))()                                                               
  yas--load-pending-jits()                                                                                                                                                                              
  yas-minor-mode(1)                                                                                                                                                                                     
  yas-minor-mode-on()                                                                                                                                                                                   
  yas-global-mode(1)                                                                                                                                                                                    
  eval-buffer(#<buffer  *load*> nil "/home/optonox/.emacs.d/init.el" nil t)  ; Reading at buffer position 1818                                                                                          
  load-with-code-conversion("/home/optonox/.emacs.d/init.el" "/home/optonox/.emacs.d/init.el" nil nil)                                                                                                  
  load("/home/optonox/.emacs.d/init.el" nil nil t)                                                                                                                                                      
  load-file("~/.emacs.d/init.el")                                                                                                                                                                       
  call-interactively(load-file record nil)                                                                                                                                                              
  command-execute(load-file record)                                                                                                                                                                     
  execute-extended-command(nil "load-file")                                                                                                                                                             
  call-interactively(execute-extended-command nil nil)                                                                                                                                                  
  command-execute(execute-extended-command)  
```

And my init file is located [HERE](https://github.com/computersarecool/dotemacs/blob/master/init.el)

I really appreciate your help.

**Comment by npostavs** [comment]
Hmm, I can't see anything bad in your init file, but it does look some css file must have been opened. Can you try adding `(debug-on-entry 'yas--schedule-jit)` to the top of your init file?

**Comment by computersarecool** [comment]
This is so strange.
Adding that right below `setq debug-on-error t` at the top of my file doesn't do anything.
Additionally, just reloading my init file after making those changes, doesn't do anything (i.e. no stack trace or debug break point) just the same error message.

(No additional output or stack trace).

I did have a prior open .css buffers though.  Closing that file made the error not appear when running my init file.

However, closing and reopening, created this error:

```
Debugger entered--entering a function:
* yas--schedule-jit(html-mode (lambda nil (yas--load-directory-1 (quote "/home/optonox/.emacs.d/snippets/html-mode") (quote html-mode))))
  yas-load-directory("~/.emacs.d/snippets" t)
  yas--load-snippet-dirs(nil)
  byte-code("\306\307\310\311 \"^X^Y^H\2031^@\n\203+^@\312\313!\203^^^@\314\315^H\"\210\2021^@\316\317\320\"\210\321\322\306\"\210\2021^@\314\323\311 \"\210\324 ^S\324 ^T\325\326^M\"\210\324 ^U\324 ^$
  yas-reload-all()
  yas--global-mode-reload-with-jit-maybe()
  run-hooks(yas-global-mode-hook yas-global-mode-on-hook)
  yas-global-mode(1)
  eval-buffer(#<buffer  *load*> nil "/home/optonox/.emacs.d/init.el" nil t)  ; Reading at buffer position 1854
  load-with-code-conversion("/home/optonox/.emacs.d/init.el" "/home/optonox/.emacs.d/init.el" t t)
  load("/home/optonox/.emacs.d/init" t t)
  #[0 "^H\205\262^@     \306=\203^Q^@\307^H\310Q\202;^@ \311=\204^^^@\307^H\312Q\202;^@\313\307\314\315#\203*^@\316\202;^@\313\307\314\317#\203:^@\320\nB^R\321\202;^@\316\322^S\323^A\322\211#\210^K\3$
  #[0 "\300 \210\302\211^Q\207" [#[0 "^H\205\262^@      \306=\203^Q^@\307^H\310Q\202;^@ \311=\204^^^@\307^H\312Q\202;^@\313\307\314\315#\203*^@\316\202;^@\313\307\314\317#\203:^@\320\nB^R\321\202;^@\$
  funcall(#[0 "\300 \210\302\211^Q\207" [#[0 "^H\205\262^@      \306=\203^Q^@\307^H\310Q\202;^@ \311=\204^^^@\307^H\312Q\202;^@\313\307\314\315#\203*^@\316\202;^@\313\307\314\317#\203:^@\320\nB^R\321$
  command-line()
  normal-top-level()
```

**Comment by computersarecool** [comment]
Wow, that was rough.
I guess that having a file loaded in a buffer and open when trying to update causes this huge problem.  Closing that buffer fixed the issue.
Thanks for your help.

**Comment by computersarecool** [comment]
@npostavs off-topic but I can't find out where you made this comment:
You mentioned, about my `init.el` file:
`(add-to-list 'auto-mode-alist '("\\.js" . js2-mode))`
Should have to trailing back slashes after the `.js`.  The original example online also had that.  For some reason though, when I take out the two trailing backslashes I get:

File mode specification error: (invalid-regexp "Trailing backslash") 

Does that make sense to you?

**Comment by npostavs** [comment]
> I guess that having a file loaded in a buffer and open when trying to update causes this huge problem.

While updating the package you mean? I'm not able to reproduce that. Can you figure out the exact sequence to trigger this?

> Should have to trailing back slashes after the .js. The original example online also had that.

You should have two backslashes followed by a `'`: `\\'`, see [`(elisp) Regexp Backslash`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Regexp-Backslash.html#index-_005c_0027-in-regexp)

**Comment by computersarecool** [comment]
I am not sure I can recreate it now that my packages have successfully been updated.
I think what was happening (what it looks like to me) was that I had an older timestamped version of Yasnippet installed (with the associated older snippets directory) in my `elpa` folder.

I also had a `.css` file open.

Then, when I updated Yasnippet to the latest version, I got that error after running the "automatic update" part of my `.init.el` file.

That part, which I do not fully understand, checks to see if all the packages are installed and updates them and then installs them if they are not already.  This <- part of the init process was where the error was happening (and then on subsequent opening of files), so perhaps that is where the issue was?

> You should have two backslashes followed by a ': \\', see (elisp) Regexp Backslash
Ahh! I was missing the trailing `'` Thank you
